python/ray/data/_internal/batcher.py

```diff
@@ -77,4 +119,205 @@ def next_batch(self) -> Block:
+            raise ValueError(
+                "Shuffle buffer capacity must be at least as large as the shuffle "
+                "buffer min size plus the batch size, but got: "
+                f"shuffle_buffer_capacity={shuffle_buffer_capacity}, "
+                f"shuffle_buffer_min_size={shuffle_buffer_min_size}, "
+                f"batch_size={batch_size}"
+            )
+        self._buffer_capacity = shuffle_buffer_capacity
+        self._buffer_min_size = shuffle_buffer_min_size
+        self._builder = DelegatingBlockBuilder()
+        self._shuffle_buffer: Block = None
+        self._shuffle_indices: List[int] = None
+        self._batch_head = 0
+        self._done_adding = False
+
+        if shuffle_seed is not None:
+            random.seed(shuffle_seed)
+
+    def add(self, block: Block):
+        """Add a block to the shuffle buffer.
+
+        Args:
+            block: Block to add to the shuffle buffer.
+        """
+        assert self.can_add(block)
+        self._builder.add_block(block)
+
+    def can_add(self, block: Block) -> bool:
+        """Whether the block can be added to the shuffle buffer.
+
+        This does not take the to-be-added block size into account when checking the
+        buffer size vs. buffer capacity, since we need to support large outlier blocks
+        and have to guard against min buffer size liveness issues.
+        """
+        return self._buffer_size() <= self._buffer_capacity and not self._done_adding
+
+    def done_adding(self) -> bool:
+        """Indicate to the batcher that no more blocks will be added to the batcher.
+
+        No more blocks should be added to the batcher after calling this.
+        """
+        self._done_adding = True
+
+    def has_any(self) -> bool:
+        """Whether this batcher has any data."""
+        return self._buffer_size() > 0
+
+    def has_batch(self) -> bool:
+        """Whether this batcher has any batches."""
+        buffer_size = self._buffer_size()
+        # If still adding blocks, ensure that removing a batch wouldn't cause the
+        # shuffle buffer to dip beneath its configured minimum size.
+        return buffer_size - self._batch_size >= self._buffer_min_size or (
+            self._done_adding and buffer_size >= self._batch_size
+        )
+
+    def _buffer_size(self) -> int:
+        """Return shuffle buffer size."""
+        buffer_size = self._builder.num_rows()
+        if self._shuffle_buffer is not None:
+            # Include the size of the concrete (materialized) shuffle buffer, adjusting
+            # for the batch head position, which also serves as a counter of the number
+            # of already-yielded rows from the current concrete shuffle buffer.
+            buffer_size += (
+                BlockAccessor.for_block(self._shuffle_buffer).num_rows()
+                - self._batch_head
+            )
+        return buffer_size
+
+    def next_batch(self) -> Block:
```

Comment:
With multiple batches, will it really be a local shuffle? If it span multiple blocks, it is still a distributed shuffle, just potentially a little smaller scale than windowed shuffle.

Reply:
They are performed on a single node, but this is because you may download multiple potentially distributed blocks onto it. I'd say it's solving distributed shuffle by turning it into non-distributed, rather than actually performing local shuffle.

Yes it's downloading them anyway, but then you have to provision more memory on the trainer node. So this may make it a GPU memory concern, which doesn't seem quite compatible with what said below about CPU/GPU constraints for two *_size parameters.

Reply:
> They are performed on a single node, but this is because you may download multiple potentially distributed blocks onto it.

But these blocks are isolated to the trainer's shard of the source dataset, while the distributed shuffle operation pulls samples from the global dataset, which is another sense in which this shuffle is "local", i.e. local to the trainer's shard.

> I'd say it's solving distributed shuffle by turning it into non-distributed, rather than actually performing local shuffle.

By that logic, Datasets has no "local" operations, e.g. batching can happen across blocks but in my opinion happens "locally" since the actual batching operation happens in the consumer's iterator. In terms of implementation, since this shuffle fundamentally fills a local buffer and pops random samples from it, I think it's local in that sense.

I think that "distributed" vs "local" might be the wrong distinction, I've mostly been thinking about this in terms of "global" vs "local", where we're talking about the quality of the random sampling and whether data is mixed across trainer shards (global data mixing). There's also the distinction of "global" vs "local" in terms of the blocks within a trainer's shard, i.e. whether we mix samples over time or if we only pick random samples from each block: this shard-local shuffle is global within the shard, but it's also not a perfectly uniform shuffle of the shard unless the memory buffer is as large as the shard.

Anyway, what are you pushing for here? That we clarify that samples are mixed from the entire shard, over time?

> Yes it's downloading them anyway, but then you have to provision more memory on the trainer node. So this may make it a GPU memory concern, which doesn't seem quite compatible with what said below about CPU/GPU constraints for two *_size parameters.

How would this be a GPU memory concern if the user needs to increase the CPU memory for the sake of shuffling capacity? The buffering happens in CPU memory so I don't see how that's a GPU memory concern.